src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala

```diff
@@ -169,9 +163,50 @@ class TaskReplaceActor(
 }
 
 object TaskReplaceActor {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
   val KillComplete = "^TASK_(ERROR|FAILED|FINISHED|LOST|KILLED)$".r
   val FailedToStart = "^TASK_(ERROR|FAILED|LOST|KILLED)$".r
 
   case object RetryKills
+
+  /** Encapsulates the logic how to get a Restart going */
+  private[upgrade] case class RestartStrategy(nrToKillImmediately: Int, maxCapacity: Int)
+
+  private[upgrade] def computeRestartStrategy(app: AppDefinition, runningTasksCount: Int): RestartStrategy = {
+    // in addition to an app definition which passed validation, we require:
+    require(app.instances > 0, s"instances must be > 0 but is ${app.instances}")
+    require(runningTasksCount >= 0, s"running task count must be >=0 but is $runningTasksCount")
+
+    val minHealthy = (app.instances * app.upgradeStrategy.minimumHealthCapacity).ceil.toInt
+    var maxCapacity = (app.instances * (1 + app.upgradeStrategy.maximumOverCapacity)).toInt
+    var nrToKillImmediately = math.max(0, runningTasksCount - minHealthy)
+
+    if (minHealthy == maxCapacity && runningTasksCount >= maxCapacity) {
+      if (app.isResident) {
+        // Kill enough tasks so that we end up with end up with one task below minHealthy.
+        // TODO: We need to do this also while restarting, since the kill could get lost.
+        nrToKillImmediately = runningTasksCount - minHealthy + 1
+        log.info(
+          "maxCapacity == minHealthy for resident app: " +
+            s"adjusting nrToKillImmediately to $nrToKillImmediately in order to prevent over-capacity for resident app"
+        )
+      }
+      else {
+        log.info(s"maxCapacity == minHealthy: Allow temporary over-capacity of one task to allow restarting")
+        maxCapacity += 1
+      }
+    }
+
+    log.info(s"For minimumHealthCapacity ${app.upgradeStrategy.minimumHealthCapacity} of ${app.id.toString} leave " +
+      s"$minHealthy tasks running, maximum capacity $maxCapacity, killing $nrToKillImmediately of " +
+      s"$runningTasksCount running tasks immediately")
+
+    assume(nrToKillImmediately >= 0, s"nrToKillImmediately must be >=0 but is $nrToKillImmediately")
```

Comment:
No, we cannot. It also depends on the number of running tasks.

E.g. if we have 3 running tasks, and a min health capacity of 4 and a max capacity of 4, we still would not need to kill a task.